nats_test.go

```diff
@@ -276,6 +276,69 @@ var testServers = []string{
 	"nats://localhost:1228",
 }
 
+func TestMaxConnectionsReconnect(t *testing.T) {
+	// Join cluster on first server
+	routes := []*url.URL{&url.URL{Scheme: "nats", Host: fmt.Sprintf("127.0.0.1:%d", TEST_PORT)}}
+
+	// Start first server
+	s1Opts := natsserver.DefaultTestOptions
+	s1Opts.Port = -1
+	s1Opts.MaxConn = 2
+	s1Opts.Cluster = server.ClusterOpts{Name: "test", Host: "127.0.0.1", Port: TEST_PORT}
+	s1Opts.Routes = routes
+	s1 := RunServerWithOptions(&s1Opts)
+	defer s1.Shutdown()
+
+	// Start second server
+	s2Opts := natsserver.DefaultTestOptions
+	s2Opts.Port = -1
+	s2Opts.MaxConn = 2
+	s2Opts.Cluster = server.ClusterOpts{Name: "test", Host: "127.0.0.1", Port: TEST_PORT + 1}
+	s2Opts.Routes = routes
+	s2 := RunServerWithOptions(&s2Opts)
+	defer s2.Shutdown()
+
+	// Only explicitly connect to first server
+	var opts = Options{
```

Comment:
`opts` here will become shared state for the two clients connections, it would be better to change into functional options instead just in case: 
```
	opts := []Option{
		MaxReconnect(2),
		ReconnectWait(10*time.Millisecond),
		Timeout(200*time.Millisecond),
	}

	// Create two connections (the current max) for first server.
	nc1, _ := nats.Connect(s1.ClientURL(), opts...)
	defer nc1.Close()
	nc1.Flush()

	nc2, _ := nats.Connect(s1.ClientURL(), opts...)
	defer nc2.Close()
	nc2.Flush()
```

Reply:
True..

Reply:
Great recommendations @kozlovic and @wallyqs .  Tips from the best.

I have made all the changes and added the commit to the PR.